### DISCUSSION TITLE
NoSuchModuleError postgres:// dialect was renamed to postgresql://

### DISCUSSION THREAD

**Discussion by transfluxus** [sqlalchemy/sqlalchemy#5799] [Usage Questions] [answered]
**Describe the bug**
cannot create engine for postgres dialect

**Expected behavior**

**To Reproduce**
SQLLAchemy version 1.4.0b1, calling create_engine   with a postgres uri: postgres:// or postgres+psycopg2://

```py
uri = f"postgres+psycopg2://{settings.POSTGRES_USER}:{settings.POSTGRES_PASSWORD.get_secret_value()}" \
      f"@{settings.POSTGRES_HOST}/{settings.POSTGRES_DB}"
try:
    engine = create_engine(uri)  # , pool_pre_ping=True)
    Base.metadata.create_all(engine)
    return engine
except OperationalError:
....
```

**Error**

```
app.db                                   INFO      postgres+psycopg2://pgman:**********@localhost/licci
Process SpawnProcess-3:
Traceback (most recent call last):
  File "/usr/lib/python3.8/multiprocessing/process.py", line 315, in _bootstrap
    self.run()
  File "/usr/lib/python3.8/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/mnt/SSD/projects/opentek_be/app/venv/lib/python3.8/site-packages/uvicorn/subprocess.py", line 61, in subprocess_started
    target(sockets=sockets)
  File "/mnt/SSD/projects/opentek_be/app/venv/lib/python3.8/site-packages/uvicorn/_impl/asyncio.py", line 47, in run
    loop.run_until_complete(self.serve(sockets=sockets))
  File "/usr/lib/python3.8/asyncio/base_events.py", line 616, in run_until_complete
    return future.result()
  File "/mnt/SSD/projects/opentek_be/app/venv/lib/python3.8/site-packages/uvicorn/_impl/asyncio.py", line 54, in serve
    config.load()
  File "/mnt/SSD/projects/opentek_be/app/venv/lib/python3.8/site-packages/uvicorn/config.py", line 306, in load
    self.loaded_app = import_from_string(self.app)
  File "/mnt/SSD/projects/opentek_be/app/venv/lib/python3.8/site-packages/uvicorn/importer.py", line 20, in import_from_string
    module = importlib.import_module(module_str)
  File "/usr/lib/python3.8/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
  File "<frozen importlib._bootstrap>", line 991, in _find_and_load
  File "<frozen importlib._bootstrap>", line 975, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 671, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 783, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "./main.py", line 19, in <module>
    from app.setup import setup_all
  File "./app/setup/__init__.py", line 8, in <module>
    from app.services.service_worker import ServiceWorker
  File "./app/services/service_worker.py", line 8, in <module>
    from app.db import Session
  File "./app/db.py", line 48, in <module>
    sessionmaker(autocommit=False, autoflush=False, bind=setup_db())
  File "./app/db.py", line 32, in setup_db
    engine = create_engine(uri, future=True) #, pool_pre_ping=True)
  File "<string>", line 2, in create_engine
  File "/mnt/SSD/projects/opentek_be/app/venv/lib/python3.8/site-packages/sqlalchemy/util/deprecations.py", line 279, in warned
    return fn(*args, **kwargs)
  File "/mnt/SSD/projects/opentek_be/app/venv/lib/python3.8/site-packages/sqlalchemy/engine/create.py", line 511, in create_engine
    entrypoint = u._get_entrypoint()
  File "/mnt/SSD/projects/opentek_be/app/venv/lib/python3.8/site-packages/sqlalchemy/engine/url.py", line 653, in _get_entrypoint
    cls = registry.load(name)
  File "/mnt/SSD/projects/opentek_be/app/venv/lib/python3.8/site-packages/sqlalchemy/util/langhelpers.py", line 302, in load
    raise exc.NoSuchModuleError(
sqlalchemy.exc.NoSuchModuleError: Can't load plugin: sqlalchemy.dialects:postgres.psycopg2
```

**Versions.**
 - OS: ubuntu 20.10
 - Python: 3.8
 - SQLAlchemy: 1.4.0b1
 - Database: postgres
 - DBAPI: psycopg2

**Have a nice day!**

**Comment by CaselIT** [comment, marked as answer]
Hi, 

As indicated in the documentation, the proper uri string is `postgresql` not `postgres`. 
See https://docs.sqlalchemy.org/en/13/dialects/postgresql.html#module-sqlalchemy.dialects.postgresql.psycopg2

**Comment by zzzeek** [comment]
"postgres" has been deprecated for many years, see https://docs.sqlalchemy.org/en/14/changelog/changelog_14.html#change-b2b94629c3428a7a3d1e7d10546de36f for removal note

**Comment by CaselIT** [comment]
Moving to discussion

**Comment by transfluxus** [comment]
great, indeed. but it was running like this for years, without warning :)

**Reply by CaselIT** [reply]
Strange, the warning is there in 1_3 (and I think since 0.x). See

https://github.com/sqlalchemy/sqlalchemy/blob/2aa511267fcb1c6e955f9fa715f78eba34f6e2d0/lib/sqlalchemy/dialects/__init__.py#L39-L42

**Comment by zzzeek** [comment]
Python doesn't show warnings by default these days and you need to have appropriate warnings flags enabled, these flags are typically turned on by test frameworks also:

```
[classic@photon3 sqlalchemy:rel_1_3]$ python -Werror
Python 3.8.3 (default, May 23 2020, 16:34:37) 
[GCC 9.3.1 20200408 (Red Hat 9.3.1-2)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from sqlalchemy import create_engine
>>> e = create_engine("postgres://scott:tiger@localhost/test")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/engine/__init__.py", line 520, in create_engine
    return strategy.create(*args, **kwargs)
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/engine/strategies.py", line 61, in create
    entrypoint = u._get_entrypoint()
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/engine/url.py", line 172, in _get_entrypoint
    cls = registry.load(name)
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/util/langhelpers.py", line 263, in load
    loader = self.auto_fn(name)
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/dialects/__init__.py", line 39, in _auto_fn
    util.warn_deprecated(
  File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/util/deprecations.py", line 23, in warn_deprecated
    warnings.warn(msg, exc.SADeprecationWarning, stacklevel=stacklevel)
sqlalchemy.exc.SADeprecationWarning: The 'postgres' dialect name has been renamed to 'postgresql'

```

**Comment by sdutton-personal** [comment]
FYI, this bit me as well, thanks for opening this as I was able to find this easier than the notes about the actual documentation.

**Reply by sdutton-personal** [reply]
I was also using `use_batch_mode=True` which had to be changed at the same time:
https://docs.sqlalchemy.org/en/13/dialects/postgresql.html#psycopg2-fast-execution-helpers
I went with:
`executemany_mode='values'`

**Comment by jeames00** [comment]
Thank you for posting this, it has been bugging me intermittently for a long time. Admittedly, in the traceback there was a reference to 'deprecations.py' in relation to the dialect statement I was using so this should have been obvious.

**Comment by dodumosu** [comment]
this is a bit of a problem, given that `postgres://` is incredibly popular. for example, i just deployed to Heroku and immediately got bitten. and Heroku doesn't allow you alter `DATABASE_URL`, at least for hobby databases

**Comment by boredland** [comment]
whoever decided to deprecate "postgres://": this is a strange decision. it maybe did cost this library 2 loc to maintain. Removing it spawned tons of issues on GitHub and Stackoverflow (for this library and tools depending on it) and forced tons of users to implement "if startswith then replace"-boilerplate.

**Reply by CaselIT** [reply]
On hindsight we should have probably kept it, since the amount of noise it generated is really not proportional with the change

The strangest thing is that a lot of people completely ignore warnings. The deprecation warning was added 13 years ago in 8fc5005dfe3eb66a46470ad8a8c7b95fc4d6bdca
We honestly did not expect such an impact from a feature that was deprecated so long ago...

**Reply by boredland** [reply]
I mean: heroku just can't change it, because it will break everything else. They could offer an additional one though, no idea why they don't. Currently I am dealing with a strange bug/setup issue in airflow where I provide a postgresql:// url but this message is thrown.

**Reply by zzzeek** [reply]
heroku had 13 years to tell us this was a problem.  they are profiting off of the work we do for free.   im sure in a few years this will have been worked out somehow, but the SQLAlchemy project is under no obligation to keep ancient names around forever.

**Reply by boredland** [reply]
Ya.. if it were only Heroku, I would agree.

**Reply by zzzeek** [reply]
hi, sorry who else is having a problem with this?  our understanding was that Heroku has some kind of hardcoded project template somewhere that they can't change.  For everyone else (i.e. those people who are not using Heroku), they merely need to go into their configs or wherever they set URLs and just change "postgres" to "postgresql", for which they have been seeing warnings regarding for up to 13 years.  Am I missing something?

**Reply by boredland** [reply]
f.ex. airflow generates the the addresses programmatically wrong. which works because they did not yet update to 1.4
![image](https://user-images.githubusercontent.com/4662748/150554499-bec18b4d-c8e1-465a-b5fd-b7c681590804.png)

**Reply by zzzeek** [reply]
post a bug with airflow?   i dont understand the problem

**Reply by zzzeek** [reply]
can't find where this happens but in airflow sourcecode they seem to know about the correct token

https://github.com/apache/airflow/blob/b5b4020d2a17185dac7ad6e4642b4bf5302eb29f/tests/cli/commands/test_info_command.py#L55

**Reply by boredland** [reply]
neither did I, would have created a PR otherwise. sure, all of this is certainly fixable. that was not my point anyway.

**Reply by zzzeek** [reply]
you're moving the goalposts.  your assertion is, " forced tons of users to implement "if startswith then replace"-boilerplate".  my assertion is, that's just users of heroku who should be getting their vendor to fix their platform.     everyone else should have changed "postgres://" to "postgresql://" years ago, not aware of any technical reason why that's a problem.

**Reply by boredland** [reply]
[mentioned](https://github.com/apache/airflow/issues/20484#issuecomment-1018628604) it in one of the upgrade preparation issues at airflow.

**Reply by boredland** [reply]
I have no numbers, but would guess that the number of python users among the heroku users (and lets not forget open-source clones like dokku which want to be compatible) may be one of the smallest fractions. and sqlalchemy is the only lib not accepting it anymore. 
dokku, however, added a variable POSTGRES_DATABASE_SCHEME that lets users define the prefix, which is kinda your point, I guess.

**Reply by zzzeek** [reply]
no way small open source projects should be constrained by large companies that dont bother to act upon deprecation warnings.   am i annoyed at them  ? absolutely.

**Reply by slyapustin** [reply]
@boredland I created PR to address that in Airflow: https://github.com/apache/airflow/pull/24569

**Comment by brianmaissy** [comment]
In case anyone found this discussion but is already using the correct URI scheme:

A similar warning can be raised by accidentally passing a `postgres_where` kwarg instead of the correct `postgresql_where` when creating a [partial index](https://docs.sqlalchemy.org/en/14/dialects/postgresql.html?highlight=postgresql_where#partial-indexes).
